cassandra/cluster.py

```diff
@@ -3521,6 +3521,18 @@ def _on_timeout(self):
     def _on_speculative_execute(self):
         self._timer = None
         if not self._event.is_set():
+
+            # PYTHON-836, the speculative queries must be after
+            # the query is sent from the main thread, otherwise the
+            # query from the main thread may raise NoHostAvailable
+            # if the _query_plan has been exhausted by the specualtive queries.
+            # This also prevents a race condition accessing the iterator.
+            # We reschedule this call until the main thread has succeeded
+            # making a query
+            if not self.attempted_hosts:
+                self.session.cluster.connection_class.create_timer(0.01, self._on_speculative_execute)
```

Comment:
Not sure what delay to set here, we have some `time.sleep(0.1)` in other parts so I figured it wouldn't be that bad.